### PR TITLE
feat(stdlib): Add new `object_from_array` function

### PR DESCRIPTION
**Pull request by bruceg**
## Summary

Adds a new `object_from_array` function as discussed in #1157 

## Change Type

- [ ] Bug fix
- [X] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?

- [ ] Yes
- [x] No

## How did you test this PR?

Unit tests are included

## Does this PR include user facing changes?

- [X] Yes. Please add a changelog fragment based on
  our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist

- [ ] Our [CONTRIBUTING.md](https://github.com/vectordotdev/vrl/blob/main/CONTRIBUTING.md) is a good starting place.
- [ ] If this PR introduces changes to [LICENSE-3rdparty.csv](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv), please
  run `dd-rust-license-tool write` and commit the changes. More details [here](https://crates.io/crates/dd-rust-license-tool).
- [ ] For new VRL functions, please also create a sibling PR in Vector to document the new function.

### CODE REVIEW

**Comment by bruceg** [comment]
Note that I plan a follow-up PR to add support for an optional `keys` parameter to avoid the need for `zip`.

Also, I don't really like the name `object_from_array` and would love a better suggestion. If you think `from_array` actually makes more sense, I can do that too.

**Comment by bruceg** [comment]
Also, this duplicates the `enum ConstOrExpr` type from `zip`. Where would that best live in common code?

File: src/stdlib/object_from_array.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        make_object(self.values.resolve(ctx)?.try_array()?)
+    }
+
+    fn type_def(&self, _state: &TypeState) -> TypeDef {
+        TypeDef::object(Collection::any())
+    }
+}
+
+#[derive(Clone, Debug)]
+enum ConstOrExpr {
```
Comment (bruceg) [inline src/stdlib/object_from_array.rs:76]: This is a straight-up clone of the construct in `zip`. Where would it make sense to move it into common code?

Comment (pront) [reply src/stdlib/object_from_array.rs:76]: Yes, we can introduce a `src/stdlib/utils.rs` (for the lack of a better name)

**Review by pront — APPROVED** [review]
Thanks!

**Pull request merged**